src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs

```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
         public override BoundNode VisitTupleCreationExpression(BoundTupleCreationExpression node)
         {
-            var rewrittenArguments = VisitList(node.Arguments);
-            return new BoundObjectCreationExpression(node.Syntax, node.ConstructorOpt, rewrittenArguments);
+            ImmutableArray<BoundExpression> rewrittenArguments = VisitList(node.Arguments);
+            return RewriteTupleCreationExpression(node, rewrittenArguments);
+        }
+
+        /// <summary>
+        /// Converts the expression for creating a tuple instance into an expression creating a ValueTuple (if short) or nested ValueTuples (if longer).
+        ///
+        /// For instance, for a long tuple we'll generate:
+        /// creationExpression(ctor=largestCtor, args=firstArgs+(nested creationExpression for remainder, with smaller ctor and next few args))
+        /// </summary>
+        private BoundNode RewriteTupleCreationExpression(BoundTupleCreationExpression node, ImmutableArray<BoundExpression> rewrittenArguments)
+        {
+            int remainder;
+            int chainLength = TupleTypeSymbol.NumberOfValueTuples(rewrittenArguments.Length, out remainder);
+
+            NamedTypeSymbol underlyingTupleType = ((TupleTypeSymbol)node.Type).UnderlyingTupleType;
+            ArrayBuilder<NamedTypeSymbol> underlyingTupleTypeChain = GetUnderlyingTypeChain(underlyingTupleType);
+
+            var currentCreation = default(BoundObjectCreationExpression);
+
+            {
+                // make a creation expression for the smallest type
+                ImmutableArray<BoundExpression> ctorArguments = ImmutableArray.Create(rewrittenArguments, (chainLength - 1) * TupleTypeSymbol.chunk7, remainder);
+
+                var ctor = (MethodSymbol)Binder.GetWellKnownTypeMember(_compilation, TupleTypeSymbol.GetTupleCtor(remainder), _diagnostics, syntax: node.Syntax);
+                if ((object)ctor == null)
+                {
+                    underlyingTupleTypeChain.Free();
+                    return node;
+                }
+
+                MethodSymbol constructor = ctor.AsMember(underlyingTupleTypeChain.Pop());
+                currentCreation = new BoundObjectCreationExpression(node.Syntax, constructor, ctorArguments);
+            }
+
+            // make successively larger creation expressions containing the previous one
+            for (int linkIndex = chainLength - 2; linkIndex >= 0; linkIndex--)
+            {
+                ImmutableArray<BoundExpression> ctorArguments = ImmutableArray.Create(rewrittenArguments, linkIndex * TupleTypeSymbol.chunk7, TupleTypeSymbol.chunk7).Add(currentCreation);
+
+                var ctor = (MethodSymbol)Binder.GetWellKnownTypeMember(_compilation, TupleTypeSymbol.GetTupleCtor(TupleTypeSymbol.chunk8), _diagnostics, syntax: node.Syntax);
+                if ((object)ctor == null)
+                {
+                    underlyingTupleTypeChain.Free();
+                    return node;
+                }
```

Comment:
The five lines above can be moved out of the loop. Surround them with an ```if``` checking that we have at least one iteration to perform and convert the loop to do-while.